connectors/utils.py

```diff
@@ -252,29 +255,65 @@ def _put(self, item):
         self._current_memsize += item[0]
         self._queue.append(item)
 
-    def mem_full(self):
-        if self.maxmemsize == 0:
-            return False
-        return self.qmemsize() >= self.maxmemsize
-
-    def qmemsize(self):
-        return self._current_memsize
+    def full(self, next_item_size=0):
+        full_by_numbers = super().full()
+        if full_by_numbers:
+            return True
+        return self._current_memsize + next_item_size >= self.maxmemsize
 
-    async def _wait_for_room(self, item):
-        item_size = get_size(item)
-        if self._current_memsize + item_size <= self.maxmemsize:
-            return item_size
+    async def _putter_timeout(self, putter):
+        """This coroutine will set the result of the putter to QueueFull when a certain timeout it reached."""
         start = time.time()
-        while self._current_memsize + item_size >= self.maxmemsize:
+        while not putter.done():
             if time.time() - start >= self.refresh_timeout:
-                raise asyncio.QueueFull()
+                putter.set_result(asyncio.QueueFull())
+                return
             logger.debug("Queue Full")
             await asyncio.sleep(self.refresh_interval)
-        return item_size
 
     async def put(self, item):
-        item_size = await self._wait_for_room(item)
-        return await super().put((item_size, item))
+        item_size = get_size(item)
+
+        # This block is taken from the original put() method but with two
+        # changes:
```

Comment:
The original put() is quite complex, have you considered explaining it quickly how it works and documentation that self._putters() is a deque used for its FIFO behavior?

Reply:
yeah, that's quite the code, it's too bad we can extend it in a simpler way. Maybe using a custom class that uses a queue internally would simplify the approach... I'll add a note